### ISSUE TITLE
hypergraph_isomorphismバグ[2.2.4]

### ISSUE THREAD

**Issue by tamanyan** [lmntal/slim#7]
## 現象

The following program generates 24 states with with --hl --nd -O0, while it generates 23 states with --hl --nd -O3:

```
init :- sequent(
    pset{c(p,var=[!X,!L,!K,!M],implied([[!X,!L],!K,[!X,!M]],[!L,!K,!M]))},
         c(p,var=[],atomic=[[],[2],[1,2]] )
                ),tset{},go.

decide@@ sequent(pset{ c($n,$m,$p), $g }, c($n1,$m1,atomic=L)),tset{$d},go :-
        unary($n),ground($m),ground($p),unary($n1),ground($m1),ground(L)|
        bchain( c($n,$m,$p), c($n1,$m1,atomic=L)),
        sequent(pset{$g,$d},c($n1,$m1,atomic=L)), tset{c($n,$m,$p)}.

to_unify2@@ bchain(c($n,var=L,implied(L1,R1)),
            c($m,var=H,atomic=R2)),unify{@d} :-
        $n=$m, ground(L1),ground(R1),ground(H),ground(L),ground(R2) |
        unify{@d,unify(L1,R2)}.

unify{
generate_eq1@@ unify([A|L],[B|R]) :- ground(A),ground(B)|
        unify(L,R),eq(A,B).
generate_eq2@@ unify([],[]) :- go .

solve2@@ eq([A|L],[B|R]),go :- sub(A,B),T=rm(L),eq(T,R),go.
solve4@@ eq(A,R),go:- hlink(A),ground(R)| sub(A,R),go.

list_check2@@ rm([A],R) :- rm(A,R).
list_check3@@ rm(A,R) :- R=A.
}.
```
## 分析

The graph isomorphism algorithm fails to identify the following two states (generated by the -O0 code), both generated by the rule "solve4":

```
18::{c(p,'='(var,[]),'='(atomic,[[],[2],[1,2]]),sequent(pset{})). tset{implied(
[[!H45,!H48],!H4c,[!H50,!H53]],[!H58,!H5b,!H5e],c(p,'='(var,[!H36,!H39,!H3c,!H3
f]))). }, unify{go. '.'(2,[],sub(!H1b)). '.'(2,[],sub(!H23)). eq([!H28,!H2c],[]
). sub(!H20,1). @5. }. @6. }
== HyperLink =============================================================
  [hl_ID]  [parent] [linked with] [num] [direct children ( inside info )]
       36        20             .     5
       39        2c             .     4
       3c        1b             .     4
       3f        23             .     4
       45        20             .     5
       48        2c             .     4
       4c        1b             .     4
       50        20             .     5
       53        23             .     4
       58        2c             .     4
       5b        1b             .     4
       5e        23             .     4
       1b      root           sub     4  3c,4c,5b.
       20      root           sub     5  28,36,45,50.
       23      root           sub     4  3f,53,5e.
       28        20             .     5
       2c      root             .     4  39,48,58.
=============================================== 4 groups, 17 elements ====
19::{c(p,'='(var,[]),'='(atomic,[[],[2],[1,2]]),sequent(pset{})). tset{implied(
[[!H45,!H48],!H4c,[!H50,!H53]],[!H58,!H5b,!H5e],c(p,'='(var,[!H36,!H39,!H3c,!H3
f]))). }, unify{go. '.'(2,[],sub(!H1e)). '.'(2,[],sub(!H23)). eq([!H28,!H2c],[]
). sub(!H1b,1). @5. }. @6. }
== HyperLink =============================================================
  [hl_ID]  [parent] [linked with] [num] [direct children ( inside info )]
       36        1b             .     5
       39        2c             .     4
       3c        23             .     4
       3f        1e             .     4
       45        1b             .     5
       48        2c             .     4
       4c        23             .     4
       50        1b             .     5
       53        1e             .     4
       58        2c             .     4
       5b        23             .     4
       5e        1e             .     4
       1b      root           sub     5  28,36,45,50.
       1e      root           sub     4  3f,53,5e.
       23      root           sub     4  3c,4c,5b.
       28        1b             .     5
       2c      root             .     4  39,48,58.
=============================================== 4 groups, 17 elements ====
```

(2014-06-06) This is due to hypergraph isomorphism checking (between a binary string encoding of a state S and a newly created state S') performed by a family of functions mem_eq_enc_\* .

The binary string encoding turned out to be correct (by dumping), but backtracking was insufficient:
There are two isomorphic subgraphs, '.'(2,[],sub(!Hxx)), inside the membrane unify. Which instance in S' corresponds to which in S should be chosen non-deterministically because !Hxx is connected to outside atoms differently.
However, the second choice is discarded one one-to-one correspondence inside the membrane unify is finished.
Then the checking of the membrane tset starts and finds that the first choice made in the membrane unify was not correct.
This problem may occur when the program uses both membranes and hyperlinks.
(2014-06-09) I've just created a flat version of the above program, which generated 23 states with both -O3 and -O0:

```
init.
init :- root(!R), pset(!P), tset(!T), uset(!U),
    c(p,var=[!X,!L,!K,!M],implied([[!X,!L],!K,[!X,!M]],[!L,!K,!M]),!P),
    sequent(!P, c(p,var=[],atomic=[[],[2],[1,2]]) ,!R), go.

decide@@ root(!R), pset(!P), tset(!T),
c($n,$m,$p,!P), sequent(!P, c($n1,$m1,atomic=L), !R), go :-
    unary($n),ground($m),ground($p),unary($n1),ground($m1),ground(L),
    hlink(P) |
    root(!R), pset(P), tset(!T),
    bchain( c($n,$m,$p), c($n1,$m1,atomic=L), !R),
    sequent(!P,c($n1,$m1,atomic=L), !R), c($n,$m,$p,!T).

to_unify2@@ root(!R), uset(!U),
bchain(c($n,var=L,implied(L1,R1)), c($m,var=H,atomic=R2), !R) :-
    $n=$m, ground(L1),ground(R1),ground(H),ground(L),ground(R2) |
    root(!R), uset(!U), unify(L1,R2,!U).

generate_eq1@@ uset(!U),
unify([A|L],[B|R],!U) :- ground(A),ground(B)|
    uset(!U), unify(L,R,!U), eq(A,B,!U).
generate_eq2@@ uset(!U), unify([],[],!U) :- uset(!U), go(!U).

solve2@@ uset(!U),
eq([A|L],[B|R],!U), go(!U) :-
    uset(!U), sub(A,B,!U), rm(L,T,!U), eq(T,R,!U), go(!U).
solve4@@ uset(!U),
eq(A,R,!U), go(!U) :- hlink(A),ground(R)|
    uset(!U), sub(A,R,!U), go(!U).

list_check2@@ uset(!U), rm([A],R,!U) :- uset(!U), rm(A,R,!U).
list_check3@@ uset(!U), rm(A,R,!U) :- uset(!U), R=A.
```
## 対応

Will take some time because the backtracking algorithm tightly follows the membrane hierarchy.
(2014-06-09) The simplest solution would probably be to handle hyperlinks as a special kind of membranes and follow them in a depth-first manner (rather than a breadth-first manner as done in the current implementation of hyperlink traversal).

**Comment by UzawaSeishiro** [comment]
2021/03/17現在のdevelopブランチ(コミットid:bb74666)でも同様のバグを確認しました.

**Comment by nikosai** [comment]
merged into #8
